### ISSUE TITLE
[dask] Dask LightGBM best practices

### ISSUE THREAD

**Issue by adfea9c0** [lightgbm-org/LightGBM#5840]
Hey! I've been experimenting a bit with Dask LightGBM and I have a couple questions. Apologies if this is not the right forum for questions like this.

1) First, what is the most recommended to pass data into LightGBM? Initially I followed the advice here [1] and persisted my data before fitting the DaskLGBMRegressor. Specifically, I would roughly use something like:
```python
df_train = dask.dataframe.read_parquet("/my/glob/*/*.parquet", columns=features+[resp], engine="pyarrow")
X_train = df_train[features].to_dask_array(lengths=True)
y_train = df_train[resp].to_dask_array(lengths=True)
X_train, y_train = X_train.persist(), y_train.persist()
_ = dask.distributed.wait([X_train, y_train])

model = lgb.DaskLGBMRegressor(**blah)
model.fit(X_train, y_train)
```
But based on early observations I'm finding this to be slower than just dropping the `persist` and `to_dask_array` calls and giving LightGBM the `read_parquet` future. I don't really understand why persisting would be helpful anyway since from what I understand, LightGBM will first do some work to rearrange the data to ensure that `X[i]` and `y[i]` are on the same worker for all `i` [2], so doing any work to load data on the workers prior seems wasted? I.e. I now run something like this:
```python
df_train = dask.dataframe.read_parquet("/my/glob/*/*.parquet", columns=features+[resp], engine="pyarrow")
model = lgb.DaskLGBMRegressor(**blah)
model.fit(df_train[features], df_train[resp])
```

2) I'm having trouble doing some kind of logging with Dask LightGBM. I'd be happy to know just how many iterations in training is, but passing in something like `callbacks=[lambda env: print(env.iteration)]` doesn't show anything in either scheduler or worker logs. Can I do some sort of logging in the dask regressor?

[1] https://github.com/jameslamb/lightgbm-dask-testing/blob/main/notebooks/demo-aws.ipynb
[2] https://www.youtube.com/watch?v=XFVcoBu1rNw

**Comment by adfea9c0** [comment]
Tangentially I'm wondering if there is some way to affect load balancing? I think either Dask or LightGBM is currently causing a very unbalanced distribution of data for my use case.

I'm loading 500 parquet files of about 1GB each onto 160 workers each with 50GB of memory. I imagine that would be plenty but some workers get very little data while others  get so much it spills to disk. How does this happen? How can I fix it ?

![image](https://user-images.githubusercontent.com/128500152/233138591-78f3615f-c244-4c31-9d48-69a9e65f22b5.png)

**Comment by matteociccozzi-ah** [comment]
@jameslamb Hey, did you ever dive deeper here? I'm experiencing some similar issues. I did a repartition call after reading the dataframe and that generates balanced partitions but not sure if that is scalable.

**Comment by jameslamb** [comment]
Thanks for the `@` and for your interest in LightGBM. No, I haven't investigated this. If you do and find some good strategies, please do post them here.

Repartitioning the DataFrame sounds like a good approach to me. `lightgbm.dask` tries to keep all of the data local to whatever worker Dask puts it on, so any imbalance is likely a result of how you create the Dask Array / DataFrame.

Here's where that's implemented, if you want to look for yourself:

https://github.com/microsoft/LightGBM/blob/82c85e4032fa12a23de341888ca545757761682e/python-package/lightgbm/dask.py#L552-L55

https://github.com/microsoft/LightGBM/blob/82c85e4032fa12a23de341888ca545757761682e/python-package/lightgbm/dask.py#L687-L690

https://github.com/microsoft/LightGBM/blob/82c85e4032fa12a23de341888ca545757761682e/python-package/lightgbm/dask.py#L697-L702

https://github.com/microsoft/LightGBM/blob/82c85e4032fa12a23de341888ca545757761682e/python-package/lightgbm/dask.py#L783-L790

**Comment by jameslamb** [comment]
Also copying my friend @ffineis , who helped implement a lot of this logic.
